pkg/apis/resource/structured/namedresources/types.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namedresources
+
+import "k8s.io/apimachinery/pkg/api/resource"
+
+// Resources is used in NodeResourceModel.
+type Resources struct {
+	Instances []Instance
+}
+
+type Instance struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	// Attributes defines the attributes of this resource instance.
+	// The name of each attribute must be unique.
+	Attributes []Attribute
+}
+
+type Attribute struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	AttributeValue
+}
+
+// AttributeValue must have one and only one field set.
+type AttributeValue struct {
+	QuantityValue    *resource.Quantity
+	BoolValue        *bool
+	IntValue         *int64
+	IntSliceValue    *IntSlice
+	StringValue      *string
```

Comment:
The protobuf encoder didn't generate usable code when this field was called just `String`, because of the conflict with the generated `String` function.

Would it be worth trying to disable that `String` function and use the simpler field names without the `Value` suffix?

Reply:
Disabling the `String` method is better, but not sufficient:

```
generated.pb.go:120:39: cannot use m (variable of type *AttributeValue) as proto.Message value in argument to xxx_messageInfo_AttributeValue.Merge: *AttributeValue does not implement proto.Message (*AttributeValue.String is a field, not a method)
```

I think we are stuck with the `Value` suffix. This is not a big deal because this primarily an API for vendors, not end-users.

Reply:
Another alternative is to use `Str` instead of `String` and drop the `Value` suffix elsewhere.

Reply:
When using `Str` as Go field name, we could still use `String` in the Kubernetes API. Will require a new `api/api-rules/violation_exceptions.list` entry for the field, but I'm leaning towards that as the best solution.

Reply:
@klueska suggested using the "Value" suffix only in the Go struct and not use it in the API:
```
StringValue string `json:"string" ...
```

That's consistent. I'll switch to that unless I hear objections soon.

Reply:
Done.

@thockin: I had to extend api/api-rules/violation_exceptions.list for this:
```patch
diff --git a/api/api-rules/violation_exceptions.list b/api/api-rules/violation_exceptions.list
index 5c08ea21b49..fec1749b318 100644
--- a/api/api-rules/violation_exceptions.list
+++ b/api/api-rules/violation_exceptions.list
@@ -52,6 +52,12 @@ API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RadosUser
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,CephFS
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,StorageOS
 API rule violation: names_match,k8s.io/api/networking/v1alpha1,ServiceCIDRSpec,CIDRs
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,BoolValue
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,IntSliceValue
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,IntValue
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,QuantityValue
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,StringSliceValue
+API rule violation: names_match,k8s.io/api/resource/v1alpha2,NamedResourcesAttributeValue,StringValue
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,Ref
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,Schema
 API rule violation: names_match,k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1,JSONSchemaProps,XEmbeddedResource
```